File: backend/match_score.py
from __future__ import annotations

import sqlite3

from . import repository as repo


def _latinize_homoglyphs(text: str) -> str:
    mapping = {
        "а": "a",
        "Ӑ": "a",
        "в": "b",
        "с": "c",
        "е": "e",
        "ё": "e",
        "н": "h",
        "к": "k",
        "м": "m",
        "о": "o",
        "п": "p",
        "р": "p",
        "т": "t",
        "х": "x",
        "у": "y",
        "і": "i",
        "ї": "i",
    }
    return "".join(mapping.get(ch, ch) for ch in text.strip().lower())


def get_scoring_points(action_name: str, outcome: str | None) -> int:
    normalized_action = _latinize_homoglyphs(action_name)
    normalized_outcome = (outcome or "").strip().lower()
    if normalized_action == "try":
        return 5
    if normalized_action.startswith("conversion"):
        return 0 if normalized_outcome == "failure" else 2
    return 0


def calculate_match_score_for_match(conn: sqlite3.Connection, match_id: int) -> dict[str, int]:
    match = repo.get_match(conn, match_id)
    if not match:
        return {"home": 0, "away": 0}

    lineups = repo.list_match_lineup(conn, match_id)
    players = repo.list_players(conn)
    events = repo.list_events_by_match(conn, match_id)
    player_to_lineup_team = {int(row["PlayerId"]): int(row["TeamId"]) for row in lineups}
    player_to_team = {int(row["Id"]): int(row["TeamId"]) for row in players}
    actions_cache: dict[int, sqlite3.Row | None] = {}

    home_team = int(match["HomeTeamId"])
    away_team = int(match["AwayTeamId"])
    score = {"home": 0, "away": 0}

    for event in events:
        action_id = event["ActionId"]
        if action_id is None:
            continue
        action_id_int = int(action_id)
        if action_id_int not in actions_cache:
            actions_cache[action_id_int] = repo.get_action(conn, action_id_int)
        action = actions_cache[action_id_int]
        if not action:
            continue
        points = get_scoring_points(str(action["Name"]), event["Outcome"])
        if points <= 0:
            continue

        team_id: int | None = None
        if event["TeamId"] is not None:
            candidate = int(event["TeamId"])
            if candidate in (home_team, away_team):
                team_id = candidate
        if team_id is None and event["PlayerId"] is not None:
            player_id = int(event["PlayerId"])
            team_id = player_to_lineup_team.get(player_id) or player_to_team.get(player_id)
            if team_id not in (home_team, away_team):
                team_id = None

        if team_id == home_team:
            score["home"] += points
        elif team_id == away_team:
            score["away"] += points

    return score

File: legacy/react-mvp/src/lib/match-score.ts
import {
  getAction,
  getMatch,
  listEventsByMatch,
  listMatchLineup,
  listPlayers,
} from '../db/repository';
import type { Action, Event, Match, MatchLineup, Player } from '../types';

export interface MatchScore {
  home: number;
  away: number;
}

/** Латинизация визуально похожих кириллических букв (Сonversion → conversion). */
function latinizeHomoglyphs(text: string): string {
  const map: Record<string, string> = {
    '\u0430': 'a',
    '\u04d0': 'a',
    '\u0432': 'b',
    '\u0441': 'c',
    '\u0435': 'e',
    '\u0451': 'e',
    '\u043d': 'h',
    '\u043a': 'k',
    '\u043c': 'm',
    '\u043e': 'o',
    '\u043f': 'p',
    '\u0440': 'p',
    '\u0442': 't',
    '\u0445': 'x',
    '\u0443': 'y',
    '\u0456': 'i',
    '\u0457': 'i',
  };

  return text
    .trim()
    .toLowerCase()
    .replace(/./g, (ch) => map[ch] ?? ch);
}

function normalizeActionName(name: string): string {
  return latinizeHomoglyphs(name);
}

function normalizeOutcome(outcome?: string): string {
  return outcome?.trim().toLowerCase() ?? '';
}

function isTryAction(name: string): boolean {
  return normalizeActionName(name) === 'try';
}

/** conversion, Conversion, conversion (opp) и т.п. */
function isConversionAction(name: string): boolean {
  return normalizeActionName(name).startsWith('conversion');
}

function isFailureOutcome(outcome?: string): boolean {
  return normalizeOutcome(outcome) === 'failure';
}

/**
 * try — 5 очков (результат не важен).
 * conversion — 2 очка при Success или если результат не указан; 0 при Failure.
 */
export function getScoringPoints(actionName: string, outcome?: string): number {
  if (isTryAction(actionName)) return 5;

  if (isConversionAction(actionName)) {
    if (isFailureOutcome(outcome)) return 0;
    return 2;
  }

  return 0;
}

/** Определяет команду-автора очка: по teamId или по игроку (состав матча → roster). */
export function resolveEventTeamId(
  event: Event,
  match: Pick<Match, 'homeTeamId' | 'awayTeamId'>,
  lineups: MatchLineup[],
  players: Player[],
): number | null {
  if (event.teamId) {
    if (event.teamId === match.homeTeamId || event.teamId === match.awayTeamId) {
      return event.teamId;
    }
  }

  if (event.playerId) {
    const lineupRow = lineups.find((l) => l.playerId === event.playerId);
    if (lineupRow) {
      if (lineupRow.teamId === match.homeTeamId || lineupRow.teamId === match.awayTeamId) {
        return lineupRow.teamId;
      }
    }

    const player = players.find((p) => p.id === event.playerId);
    if (player) {
      if (player.teamId === match.homeTeamId || player.teamId === match.awayTeamId) {
        return player.teamId;
      }
    }
  }

  return null;
}

export function calculateMatchScore(
  events: Event[],
  actions: Action[],
  match: Pick<Match, 'homeTeamId' | 'awayTeamId'>,
  lineups: MatchLineup[],
  players: Player[],
): MatchScore {
  const actionById = new Map(actions.filter((a) => a.id).map((a) => [a.id!, a]));
  const score: MatchScore = { home: 0, away: 0 };

  for (const event of events) {
    if (!event.actionId) continue;

    const action = actionById.get(event.actionId);
    if (!action) continue;

    const points = getScoringPoints(action.name, event.outcome);
    if (points === 0) continue;

    const teamId = resolveEventTeamId(event, match, lineups, players);
    if (!teamId) continue;

    if (teamId === match.homeTeamId) score.home += points;
    else if (teamId === match.awayTeamId) score.away += points;
  }

  return score;
}

export async function calculateMatchScoreForMatch(matchId: number): Promise<MatchScore | null> {
  const match = await getMatch(matchId);
  if (!match?.id) return null;

  const [events, lineups, players] = await Promise.all([
    listEventsByMatch(matchId),
    listMatchLineup(matchId),
    listPlayers(),
  ]);

  const actionIds = [...new Set(events.map((e) => e.actionId).filter((id): id is number => id != null))];
  const actions = (await Promise.all(actionIds.map((id) => getAction(id)))).filter(
    (a): a is Action => a != null,
  );

  return calculateMatchScore(events, actions, match, lineups, players);
}
